api/webrtc/webrtc.go

```diff
@@ -0,0 +1,380 @@
+			c.OnSdp <- encodedDescr
+		}
+	})
+
+	c.pc.OnDataChannel(func(dc *webrtc.DataChannel) {
+		dc.OnOpen(func() {
+			c.dc = dc
+
+			c.mutex.Lock()
+			c.isConnected = true
+			c.mutex.Unlock()
+
+			if NewConnection != nil {
+				go NewConnection(c, nil)
+			}
+		})
+
+		dc.OnMessage(func(msg webrtc.DataChannelMessage) {
+			var dcmsg DataChannelMessage
+			if msg.IsString {
+				dcmsg.messageType = TextMessage
+				if string(msg.Data) == PingData {
+					dc.SendText(PongData)
+					return
+				} else if string(msg.Data) == PongData {
+					if c.pongHandler != nil {
+						c.pongHandler(PongData)
+					} else {
+						log.Info("Pong handler not set")
+					}
+					return
+				}
+			} else {
+				dcmsg.messageType = BinaryMessage
+			}
+			dcmsg.data = msg.Data
+			c.OnMessage <- dcmsg
+		})
+
+		dc.OnClose(func() {
+			c.mutex.Lock()
+			c.isConnected = false
+			c.mutex.Unlock()
+		})
+
+		dc.OnError(func(err error) {
+			log.Errorf("Data Channel %v error: %s\n", dc.Label(), err.Error())
+		})
+	})
+
+	return nil
+}
+
+func (c *Peer) IsConnected() bool {
+	c.mutex.RLock()
+	defer c.mutex.RUnlock()
+	return c.isConnected
+}
+
+func (c *Peer) SetRemoteSdp(sdp string) error {
+	answer := webrtc.SessionDescription{}
+	err := Decode(sdp, &answer)
+	if err != nil {
+		return err
+	}
+	return c.pc.SetRemoteDescription(answer)
+}
+
+func (c *Peer) WriteMessage(messageType int, data []byte) (err error) {
+	if c.dc == nil {
+		return fmt.Errorf("DataChannel not available")
+	}
+	if data == nil {
+		if messageType == PingMessage {
+			return c.dc.SendText(PingData)
+		} else if messageType == PongMessage {
+			return c.dc.SendText(PongData)
+		}
+		return nil
+	}
+
+	writeResult := make(chan error)
+	go func() {
+		if messageType == TextMessage {
+			err = c.dc.SendText(string(data))
+		} else {
+			err = c.dc.Send(data)
+		}
+
+		writeResult <- err
+	}()
+
+	var dur time.Duration
+	switch {
+	case c.writeDeadline.After(time.Now()):
+		dur = time.Until(c.writeDeadline)
+		select {
+		case err := <-writeResult:
+			return err
+		case <-time.After(dur):
+			return os.ErrDeadlineExceeded
+		}
+
+	case c.readDeadline.Before(time.Now()):
+	case c.readDeadline == time.Now():
+		return os.ErrDeadlineExceeded
+	}
+
+	return <-writeResult
+}
+
+// WriteJSON writes the JSON encoding of v as a message.
+func (c *Peer) WriteJSON(v interface{}) error {
+	b, err := json.Marshal(v)
+	if err != nil {
+		return err
+	}
+	return c.WriteMessage(TextMessage, b)
+}
+
+func (c *Peer) ReadMessage() (messageType int, data []byte, err error) {
+	switch {
+	case c.readDeadline.After(time.Now()):
+		for {
+			oldReadDeadline := c.readDeadline
+			dur := time.Until(c.readDeadline)
+			select {
+			case msg := <-c.OnMessage:
+				return msg.messageType, msg.data, nil
+			case <-time.After(dur):
+				if c.readDeadline.After(oldReadDeadline) {
+					continue
+				}
+				return UnknownMessage, nil, os.ErrDeadlineExceeded
+			}
+		}
+
+	case c.readDeadline.Before(time.Now()):
```

Comment:
golang switch won't automatically fallthrough. You need to call `fallthrough` explicitly

Reply:
Yes, modified